src/connection/in_channel.rs

```diff
@@ -0,0 +1,144 @@
+use std::{collections::HashMap, sync::Arc};
+
+use tokio::sync::{broadcast, mpsc, Mutex};
+
+use crate::graph::node::NodeId;
+
+use super::information_packet::Content;
+
+/// # Input Channels
+/// A hash-table mapping `NodeId` to `InChannel`. In **Dagrs**, each `Node` stores input
+/// channels in this map, enabling `Node` to receive information packets from other `Node`s.
+/// ## Implementaions
+/// - `blocking_recv_from`: call `blocking_recv` of the receiver by the given `NodeId`. Returns `Ok(Content)`
+/// if message received; returns `Err(RecvErr)` if the given `NodeId` is invalid, no message is available to recv,
+/// or err occurs.
+/// - `recv_from`: receives the next value for this receiver by the given `NodeId` asynchronously. Returns `Ok(Content)`
+/// if message received; returns `Err(RecvErr)` if the given `NodeId` is invalid, or no message is available to recv,
+/// or err occurs.
+/// - `close`: close the channel by the given `NodeId`, and remove the channel in this map.
+#[derive(Default)]
+pub struct InChannels(pub HashMap<NodeId, Arc<Mutex<InChannel>>>);
+
+impl InChannels {
+    /// Call `blocking_recv` of the receiver of the given `NodeId`. Returns `Ok(Content)`
+    /// if message received; returns `Err(RecvErr)` if the given `NodeId` is invalid, no message is available to recv,
+    /// or err occurs.
```

Comment:
I prefer your suggestion to modify the document. I will update the documentation here in my next commit.

For the err type, we want users to receive messages from other nodes directly through `InChannels`, without caring what `InChannel` is. That's why I defined the methods implemented for `InChannel` as **private** items. Maybe we can define `InChannel` as a private item too.

So I think we don't have to create a separate Err type for `InChannel`, because on the user side they should not know the existence of `InChannel`.